### PR TITLE
feat: build graph endpoint for pipeline visualization

### PR DESCRIPTION
**Pull request by plyr4**
meant to work with ui changes in https://github.com/go-vela/ui/pull/731

the new endpoint aggregates resources for a build and generates a [directed graph](https://graphviz.org/Gallery/directed/) consisting of nodes and edges. 

- **nodes** and **edges** are constructed using stages and their `needs`
    - if a pipeline has no stages then each step is treated as a single-step stage
- service edges are connected sequentially to produce a line subgraph at the top of the main graph 

there are 3 node groups organized as [subgraph clusters](https://graphviz.org/Gallery/directed/cluster.html) for visual effect.
- services
- built-in (init and clone)
- pipeline (everything else)

closes https://github.com/go-vela/community/issues/71

### CODE REVIEW

File: api/build/graph.go [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+
+	// parse and compile the pipeline configuration file
+	p, _, err := compiler.FromContext(c).
+		Duplicate().
+		WithBuild(b).
+		WithFiles(files).
+		WithMetadata(m).
+		WithRepo(r).
+		WithUser(u).
+		Compile(config)
```
Comment (ecrupper) [inline api/build/graph.go:267]:
```suggestion
	// parse and compile the pipeline configuration file
	p, _, err := compiler.FromContext(c).
		Duplicate().
		WithBuild(b).
		WithFiles(files).
		WithCommit(b.GetCommit())
		WithMetadata(m).
		WithRepo(r).
		WithUser(u).
		Compile(config)
```

Will be necessary for `file` type templates.

Ref: https://github.com/go-vela/server/blob/4eec5e9b03ab7e7004bfae08fc558eb2b27d87c5/compiler/native/expand.go#L289

**Review by ecrupper — CHANGES_REQUESTED** [review]
Just a real quick thing I noticed

File: api/build/graph.go (hunk 1 above)
Comment (plyr4) [reply api/build/graph.go]: oo updated, thank you

File: api/build/graph.go [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+		cluster := PipelineCluster
+
+		// override the cluster id for built-in nodes
+		// this allows for better layout control because all stages need 'clone'
+		if stage.Name == "init" {
+			cluster = BuiltInCluster
+		}
+		if stage.Name == "clone" {
+			cluster = BuiltInCluster
+		}
```
[the remark this answers is not in the document]
Comment (plyr4) [reply api/build/graph.go:395]: i know 😓 add it to the list!

File: api/build/graph.go [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+	}
+
+	// skip the build if only the init or clone steps are found
+	skip := SkipEmptyBuild(p)
+	if skip != "" {
+		c.JSON(http.StatusOK, skip)
+		return
+	}
+
+	logger.Info("generating build graph")
```
[the remark this answers is not in the document]
Comment (plyr4) [reply api/build/graph.go:291]: yeah id appreciate a better way to aggregate all resources for a build. not scoping that for this PR but i agree we should pursue it

File: api/build/graph.go [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+					Destination: sourceNode.ID + 1,
+					Status:      sourceNode.Status,
+				}
+				edges = append(edges, edge)
+			}
+		}
+	}
+
+	// for loop over edges, and collapse same parent edge
+	// todo: move this check above the processing ?
```
[the remark this answers is not in the document]
Comment (plyr4) [reply api/build/graph.go:502]: ill do that, im going to move the complexity check above and add a constant

[the remark this answers is not in the document]
Comment (plyr4) [reply api/build/graph.go]: i suppose we could make that a server configuration, but, i dont feel its necessary

File: .gitignore [hunk 5]
```diff
@@ -48,6 +48,7 @@ api-spec.json
 # Edit at https://www.toptal.com/developers/gitignore?templates=visualstudiocode
 
 ### VisualStudioCode ###
+.vscode/
```
Comment (wass3r) [inline .gitignore:51]: is this needed?

File: api/build/graph.go [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+		"org":   o,
+		"repo":  r.GetName(),
+		"user":  u.GetName(),
+	})
+
+	baseErr := "unable to retrieve graph"
+
+	// update engine logger with API metadata
+	//
+	// https://pkg.go.dev/github.com/sirupsen/logrus?tab=doc#Entry.WithFields
```
Comment (wass3r) [inline api/build/graph.go:148]: i think this comment is meant for L137?

File: api/build/graph.go [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+
+		// assume no more pages exist if under 100 results are returned
+		//
+		// nolint: gomnd // ignore magic number
+		if len(stepsPart) < 100 {
+			page = 0
+		} else {
+			page++
+		}
+	}
```
Comment (wass3r) [inline api/build/graph.go:177]: is there a reason to not utilize the returned, but currently ignored, count to determine additional pages? it should avoid edge cases.

same for services below.

File: api/build/graph.go [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+//   '200':
+//     description: Successfully retrieved graph for the build
+//     schema:
+//       type: array
+//       items:
+//         "$ref": "#/definitions/Graph"
+//   '500':
+//     description: Unable to retrieve graph for the build
+//     schema:
+//       "$ref": "#/definitions/Error"
```
Comment (wass3r) [inline api/build/graph.go:123]: I believe this endpoint also can return 401 and 404?

File: api/build/graph.go [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+
+			return
+		}
+
+		// add page of steps to list steps
+		steps = append(steps, stepsPart...)
+
+		// assume no more pages exist if under 100 results are returned
+		//
+		// nolint: gomnd // ignore magic number
```
Comment (wass3r) [inline api/build/graph.go:171]: linter typically yells at these to for not being `//nolint:` vs `// nolint:` 🤔

**Review by wass3r — CHANGES_REQUESTED** [review]

File: api/build/graph.go [hunk 10]
```diff
@@ -0,0 +1,577 @@
+// Copyright (c) 2021 Target Brands, Inc. All rights reserved.
+//
+// Use of this source code is governed by the LICENSE file in this repository.
```
[the remark this answers is not in the document]
Comment (wass3r) [reply api/build/graph.go:3]: found the reason in the GitHub Actions run:

>  level=error msg="Running error: context loading failed: failed to load packages: timed out to load packages: context deadline exceeded"
>  level=error msg="Timeout exceeded: try increasing it by passing --timeout option"

File: .gitignore (hunk 5 above)
Comment (plyr4) [reply .gitignore]: probably not, likely left there from before i added the official one. removed!

File: api/build/graph.go (hunk 6 above)
Comment (plyr4) [reply api/build/graph.go:149]: yup! hehe

File: api/build/graph.go (hunk 7 above)
Comment (plyr4) [reply api/build/graph.go:178]: nope, i'll fix that, thank you

File: api/build/graph.go (hunk 9 above)
Comment (plyr4) [reply api/build/graph.go]: you can tell how long ago i copy pasted the code 😅 i removed them

File: api/build/graph.go (hunk 8 above)
Comment (plyr4) [reply api/build/graph.go:124]: true! adjusting

File: api/build/graph.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (plyr4) [reply api/build/graph.go]: just cleaned up the linter feedback and adjusted headers. thank you 🙏

File: api/build/graph.go [hunk 11]
```diff
@@ -0,0 +1,594 @@
...
+			util.HandleError(c, http.StatusNotFound, retErr)
+
+			return
+		}
+
+		// add page of steps to list steps
+		steps = append(steps, stepsPart...)
+
+		// assume no more pages exist if under 100 results are returned
+		if stepsCount < 100 {
```
Comment (wass3rw3rk) [inline api/build/graph.go:179]: should we use `perPage` instead of `100` here? same for services.

Comment (plyr4) [reply api/build/graph.go:179]: good call, easy!

**Review by wass3r — DISMISSED** [review]
let's go to the moon! 🌑

**Comment by wass3r** [comment]
one parting thought, should we try and fetch pipeline first and make calls to get steps/services dependent on whether pipeline has steps/services? might save some db calls, esp for services?

**Review by ecrupper — DISMISSED** [review]
Very cool. Opened an issue related to the discussion @cognifloyd and @plyr4 had in this PR: https://github.com/go-vela/community/issues/874

**Review by timhuynh94 — APPROVED** [review]

**Review by timhuynh94 — APPROVED** [review]

**Review by wass3r — DISMISSED** [review]
nice :D

**Review by cognifloyd — APPROVED** [review]

**Review by ecrupper — APPROVED** [review]
Nice

**Review by claire1618 — APPROVED** [review]
I think it looks good!

**Review by wass3rw3rk — APPROVED** [review]

**Review by timhuynh94 — APPROVED** [review]

**Pull request merged**